packages/babel-generator/src/generators/types.ts

```diff
@@ -185,45 +190,86 @@ export function NullLiteral(this: Printer) {
 }
 
 export function NumericLiteral(this: Printer, node: t.NumericLiteral) {
-  const raw = this.getPossibleRaw(node);
   const opts = this.format.jsescOption;
   const value = node.value;
-  const str = value + "";
   if (opts.numbers) {
     this.number(jsesc(value, opts), value);
-  } else if (raw == null) {
-    this.number(str, value); // normalize
+    return;
+  }
+  const raw = this.getPossibleRaw(node);
+  if (raw == null) {
+    this.number(value + "", value); // normalize
   } else if (this.format.minified) {
+    const str = value + "";
     this.number(raw.length < str.length ? raw : str, value);
   } else {
     this.number(raw, value);
   }
 }
 
-export function StringLiteral(this: Printer, node: t.StringLiteral) {
-  const raw = this.getPossibleRaw(node);
-  if (!this.format.minified && raw !== undefined) {
-    this.token(raw);
+const escapeCodeMap = new Map([
+  [34, "&quot;"],
+  [38, "&amp;"],
+  [60, "&lt;"],
+  [62, "&gt;"],
+]);
+const escapeReplacer = /["&<>]/g;
+
+// From https://www.npmjs.com/package/entities License: MIT
+function escape(data: string) {
+  let match;
+  let lastIdx = 0;
+  let result = "";
+  while ((match = escapeReplacer.exec(data))) {
+    if (lastIdx !== match.index) {
+      result += data.substring(lastIdx, match.index);
+    }
+    // We know that this character will be in the map.
+    result += escapeCodeMap.get(match[0].charCodeAt(0));
+    // Every match will be of length 1
+    lastIdx = match.index + 1;
+  }
+  return result + data.substring(lastIdx);
+}
+
+export function StringLiteral(
+  this: Printer,
+  node: t.StringLiteral,
+  parent: NodePath<
+    t.FunctionDeclaration | t.FunctionExpression | t.TSDeclareFunction
+  >["parent"],
+) {
+  const inJsx = isJSXAttribute(parent);
+
+  const raw =
+    !this.format.minified &&
+    inJsx == !!node.extra?.jsx &&
+    this.getPossibleRaw(node);
+
+  if (typeof raw === "string") {
+    this.token(raw, inJsx);
     return;
   }
 
-  const val = jsesc(node.value, this.format.jsescOption);
+  const val = inJsx
+    ? `"${escape(node.value)}"`
+    : jsesc(node.value, this.format.jsescOption);
```

Comment:
Could we not handle HTML entities here & while parsing at all, and instead handle them in the JSX transform? 🤔 

i.e. `<X y="a &lt; b" />` is just a normal string according to the parser/generator, and when transforming it to plain JS the plugin will unescape HTML entities and other JSX weirdnesses. By  doing so, we need to handle JSX encoding just in one place (the plugin) rather than twice (the parser and the generator).

Reply:
No. A string literal must be escaped differently for the different contexts. In JS, `\` must be escaped as `\\`, and in JSX, it must not. In JS, `"` must be escaped as `\"`, and in JSX, it must be escaped as `&quot;`. These differences must be handled by the parser and generator to support users of the parser and generator that transplant a string literal node from one context to another. And this doesn’t happen in just one plugin; for example, the React plugin transforms JSX differently from the Solid plugin. See #15472.

Reply:
That would be a huge breaking change, meaning escaping would have to be handled manually by plugins and users. Given that it's not difficult for us to support this, and we've supported it in the past, there are some users downstream who need it, so I'd be inclined to leave it as is.
And even so, we still need to check whether a `StringLiteral` is in jsx or js. :)